File: common/src/main/java/juuxel/adorn/block/CarpetedBlock.java
package juuxel.adorn.block;

import juuxel.adorn.block.property.OptionalProperty;
import juuxel.adorn.util.Dyes;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.RandomSource;
import net.minecraft.world.item.DyeColor;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.ScheduledTickAccess;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.WoolCarpetBlock;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.world.level.storage.loot.LootParams;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jspecify.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public abstract class CarpetedBlock extends SeatBlock {
    public static final OptionalProperty<DyeColor> CARPET = new OptionalProperty<>(EnumProperty.create("carpet", DyeColor.class, Dyes.ALL_DYES));
    public static final VoxelShape CARPET_SHAPE = box(0.0, 0.0, 0.0, 16.0, 1.0, 16.0);

    public CarpetedBlock(Properties settings) {
        super(settings);

        if (isCarpetingEnabled()) {
            registerDefaultState(defaultBlockState().setValue(CARPET, CARPET.getNone()));
        }
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
        if (isCarpetingEnabled()) builder.add(CARPET);
    }

    @Override
    public @Nullable BlockState getStateForPlacement(BlockPlaceContext ctx) {
        if (isCarpetingEnabled()) {
            return super.getStateForPlacement(ctx).setValue(CARPET, CARPET.wrapOrNone(getCarpetColor(ctx)));
        }

        return super.getStateForPlacement(ctx);
    }

    @Override
    public void tick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
        if (!isCarpetingEnabled()) return;
        var carpet = state.getValue(CARPET);
        if (carpet.isPresent()) {
            var carpetBlock = getCarpetBlock(carpet.value());
            if (!carpetBlock.defaultBlockState().canSurvive(world, pos)) {
                carpetBlock.playerWillDestroy(world, pos, state, null);
                dropResources(carpetBlock.defaultBlockState(), world, pos);
                world.setBlockAndUpdate(pos, state.setValue(CARPET, CARPET.getNone()));
            }
        }
    }

    @Override
    protected BlockState updateShape(BlockState state, LevelReader world, ScheduledTickAccess tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, RandomSource random) {
        if (isCarpetingEnabled()) {
            var carpet = state.getValue(CARPET);
            if (carpet.isPresent() && !getCarpetBlock(carpet.value()).defaultBlockState().canSurvive(world, pos)) {
                tickView.scheduleTick(pos, this, 1);
            }
        }

        return super.updateShape(state, world, tickView, pos, direction, neighborPos, neighborState, random);
    }

    @Override
    protected List<ItemStack> getDrops(BlockState state, LootParams.Builder builder) {
        if (isCarpetingEnabled() && state.getValue(CARPET).isPresent()) {
            var stacks = new ArrayList<>(super.getDrops(state, builder));
            stacks.addAll(getCarpetBlock(state.getValue(CARPET).value()).defaultBlockState().getDrops(builder));
            return stacks;
        }

        return super.getDrops(state, builder);
    }

    public boolean isCarpetingEnabled() {
        return true;
    }

    public boolean canStateBeCarpeted(BlockState state) {
        return isCarpetingEnabled() && state.getValue(CARPET) == CARPET.getNone();
    }

    private static @Nullable DyeColor getCarpetColor(BlockPlaceContext context) {
        var block = context.getLevel().getBlockState(context.getClickedPos()).getBlock();
        return block instanceof WoolCarpetBlock carpet ? carpet.getColor() : null;
    }

    private static Block getCarpetBlock(DyeColor color) {
        return Blocks.CARPET.pick(color);
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/CarpetedBlock.kt
@file:Suppress("DEPRECATION")
package juuxel.adorn.block

import juuxel.adorn.block.property.OptionalProperty
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.Blocks
import net.minecraft.block.DyedCarpetBlock
import net.minecraft.item.ItemPlacementContext
import net.minecraft.item.ItemStack
import net.minecraft.loot.context.LootContext
import net.minecraft.server.world.ServerWorld
import net.minecraft.state.StateManager
import net.minecraft.state.property.EnumProperty
import net.minecraft.util.DyeColor
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.WorldAccess
import java.util.Random

abstract class CarpetedBlock(settings: Settings) : SeatBlock(settings) {
    init {
        if (isCarpetingEnabled()) {
            defaultState = defaultState.with(CARPET, CARPET.none)
        }
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        super.appendProperties(builder)
        if (isCarpetingEnabled()) builder.add(CARPET)
    }

    override fun getPlacementState(context: ItemPlacementContext) =
        if (isCarpetingEnabled())
            super.getPlacementState(context)!!.with(CARPET, getCarpetColor(context)?.let(CARPET::wrap) ?: CARPET.none)
        else
            super.getPlacementState(context)

    override fun scheduledTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random?) {
        if (!isCarpetingEnabled()) return
        val carpet = state[CARPET]
        if (carpet.isPresent) {
            val carpetBlock = COLORS_TO_BLOCKS[carpet.value] ?: error("Unknown carpet state: $carpet")
            if (!carpetBlock.defaultState.canPlaceAt(world, pos)) {
                carpetBlock.onBreak(world, pos, state, null)
                dropStacks(carpetBlock.defaultState, world, pos)
                world.setBlockState(pos, state.with(CARPET, CARPET.none))
            }
        }
    }

    override fun getStateForNeighborUpdate(
        state: BlockState, direction: Direction, neighborState: BlockState,
        world: WorldAccess, pos: BlockPos, neighborPos: BlockPos
    ): BlockState {
        if (isCarpetingEnabled()) {
            val carpet = state[CARPET]
            if (carpet.isPresent && !COLORS_TO_BLOCKS[carpet.value]!!.defaultState.canPlaceAt(world, pos))
                world.blockTickScheduler.schedule(pos, this, 1)
        }

        return super.getStateForNeighborUpdate(state, direction, neighborState, world, pos, neighborPos)
    }

    override fun getDroppedStacks(state: BlockState, builder: LootContext.Builder): List<ItemStack> =
        if (isCarpetingEnabled()) {
            super.getDroppedStacks(state, builder) +
                (COLORS_TO_BLOCKS[state[CARPET].value]?.defaultState?.getDroppedStacks(builder) ?: emptyList())
        } else {
            super.getDroppedStacks(state, builder)
        }

    open fun isCarpetingEnabled() = true

    open fun canStateBeCarpeted(state: BlockState) =
        isCarpetingEnabled() && state[CARPET] == CARPET.none

    companion object {
        val CARPET = OptionalProperty(EnumProperty.of("carpet", DyeColor::class.java))
        val CARPET_SHAPE = createCuboidShape(0.0, 0.0, 0.0, 16.0, 1.0, 16.0)
        val COLORS_TO_BLOCKS: Map<DyeColor, Block> = mapOf(
            DyeColor.WHITE to Blocks.WHITE_CARPET,
            DyeColor.ORANGE to Blocks.ORANGE_CARPET,
            DyeColor.MAGENTA to Blocks.MAGENTA_CARPET,
            DyeColor.LIGHT_BLUE to Blocks.LIGHT_BLUE_CARPET,
            DyeColor.YELLOW to Blocks.YELLOW_CARPET,
            DyeColor.LIME to Blocks.LIME_CARPET,
            DyeColor.PINK to Blocks.PINK_CARPET,
            DyeColor.GRAY to Blocks.GRAY_CARPET,
            DyeColor.LIGHT_GRAY to Blocks.LIGHT_GRAY_CARPET,
            DyeColor.CYAN to Blocks.CYAN_CARPET,
            DyeColor.PURPLE to Blocks.PURPLE_CARPET,
            DyeColor.BLUE to Blocks.BLUE_CARPET,
            DyeColor.BROWN to Blocks.BROWN_CARPET,
            DyeColor.GREEN to Blocks.GREEN_CARPET,
            DyeColor.RED to Blocks.RED_CARPET,
            DyeColor.BLACK to Blocks.BLACK_CARPET
        )

        private fun getCarpetColor(context: ItemPlacementContext): DyeColor? =
            when (val block = context.world.getBlockState(context.blockPos).block) {
                is DyedCarpetBlock -> block.dyeColor
                else -> null
            }
    }
}
